sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -422,6 +424,108 @@ class Analyzer(catalog: Catalog,
+        val transformedConds = conditions.transform {
+          case In(_, Seq(SubqueryExpression(_))) => Literal(true)
+        }
+        subqueryExprs match {
+          case Seq() => filter // No subqueries.
+          case Seq((exp, subquery)) =>
+            createLeftSemiJoin(
+              child,
+              exp,
+              subquery,
+              transformedConds)
+          case _ =>
+            throw new TreeNodeException(filter, "Only one SubQuery expression is supported.")
+        }
+    }
+
+    /**
+     * Create LeftSemi join with parent query to the subquery which is mentioned in 'IN' predicate
+     * And combine the subquery conditions and parent query conditions.
+     */ 
+    def createLeftSemiJoin(left: LogicalPlan,
+        value: Expression,
+        subquery: LogicalPlan,
+        parentConds: Expression) : LogicalPlan = {
+      val (transformedPlan, subqueryConds) = transformAndGetConditions(value, subquery)
+      // Add both parent query conditions and subquery conditions as join conditions
+      val allPredicates = And(parentConds, subqueryConds)
+      Join(left, transformedPlan, LeftSemi, Some(allPredicates))
+    }
+
+    /**
+     * Transform the subquery LogicalPlan and add the expressions which are used as filters to the
+     * projection. And also return filter conditions used in subquery
+     */
+    def transformAndGetConditions(value: Expression,
+          subquery: LogicalPlan): (LogicalPlan, Expression) = {
+      val expr = new scala.collection.mutable.ArrayBuffer[Expression]()
+      // TODO : we only decorelate subqueries in very specific cases like the cases mentioned above
+      // in documentation. The more complex queries like using of subqueries inside subqueries can 
+      // be supported in future.
+      val transformedPlan = subquery transform {
+        case project @ Project(projectList, f @ Filter(condition, child)) =>
+          // Don't support more than one item in select list of subquery
+          if(projectList.size > 1) {
+            throw new TreeNodeException(
+                project,
+                "SubQuery can contain only one item in Select List")
+          }
+          val resolvedChild = ResolveRelations(child)
+          // Add the expressions to the projections which are used as filters in subquery
+          val toBeAddedExprs = f.references.filter{a =>
+            resolvedChild.resolve(a.name, resolver) != None && !project.outputSet.contains(a)}
+          val nameToExprMap = collection.mutable.Map[String, Alias]()
+          // Create aliases for all projection expressions.
+          val witAliases = (projectList ++ toBeAddedExprs).zipWithIndex.map {
+            case (exp, index) => 
+              nameToExprMap.put(exp.name, Alias(exp, s"sqc$index")())
+              Alias(exp, s"sqc$index")()
+          }
+          // Replace the condition column names with alias names.
+          val transformedConds = condition.transform {
```

Comment:
I am not so sure why you cares about the subquery condition, as in Hive wiki
```
As of Hive 0.13 some types of subqueries are supported in the WHERE clause. Those are queries where the result of the query can be treated as a constant for IN and NOT IN statements (called uncorrelated subqueries because the subquery does not reference columns from the parent query):
```